### ISSUE TITLE
Its very very slow

### ISSUE THREAD

**Issue by ffoxD** [thekevinscott/UpscalerJS#14]
I used www.upscaler.ai to upscale a bunch of photos, but its extremely slow... I originally thought it got stuck at 0% but it does do progress. I got to 12% as I'm writiing this, i will update if it finishes... It took a very long time tho. A Linux user, tried on Firefox and Chromium.

**Comment by ffoxD** [comment]
Okk it finally did it, but.... Please improve performance. It's horribly slow.

**Comment by ffoxD** [comment]
Maybe pixel-art sized images make i run extremely slow? The smaller is the image, the more it takes.... In fact it just went from 70% to 40% for some reason, its weird

**Comment by dziku86** [comment]
It depends on your computer processor power capability.  In Chrome press `SHIFT` + `ESC` to open task manager and see how much CPU does it use.

**Comment by danwexler** [comment]
Also much faster with a larger GPU, since it leverages TensorFlow's WebGL back-end. I use it for low-resolution images, say 128x128 all the time. Upscaling those takes about 8 seconds on my 2018 MBP. For more information on performance, I suggest looking into TensorFlow, since that's the computing engine. Much less can be done within the UpscalerJS code to improve performance.

**Comment by thekevinscott** [comment]
Agree with the advice others have posted - it's largely limited by TFJS and your running hardware.

Some things to check:

- [That you're warming up your model](https://thekevinscott.github.io/UpscalerJS/#/?id=warmup) which can speed up the initial inference
- That you're [implementing web workers](https://github.com/thekevinscott/UpscalerJS/tree/master/examples/webworker)
- That you're [setting patch sizes](https://github.com/thekevinscott/UpscalerJS/tree/master/examples/patch-sizes) (which can increase total time, but improve UI performance)

Beyond those performance improvements, it's largely up to your running hardware and the TFJS backend.

---

One thing I strongly suspect will [improve model performance is pruning, which I discuss a bit here](https://thekevinscott.com/super-resolution-with-js/#pull-your-weight-fine-neurons).

I've read reports that as much as 90% of a model's weights can be pruned (e.g., a model can be reduced to 10% of its size). Even a model 50% of the size of the ones we have would see drastic performance improvements.

Unfortunately, pruning needs to happen at train-time, which means 1) modifying the training code to support it, and 2) re-training the model from scratch. Not impossible but I haven't time to do it.

If any enterprising developers feel like exploring this, the [Python code we're using to train the code is here](https://github.com/idealo/image-super-resolution) and the [dataset we're using to train is Div2k](https://data.vision.ee.ethz.ch/cvl/DIV2K/). That Python code would need to be modified to support pruning ([example here](https://www.tensorflow.org/model_optimization/guide/pruning/pruning_with_keras)).

Unfortunately I don't know _how_ much pruning will help in this case; it's sort of a "try-it-and-see" thing. But it's the last major potential performance improvement that has yet to be explored yet in this library.
